# WWTPs-Prediction-Clustering-Generation
Prediction, Generation of WWTPs microbiome community structures and Clustering of WWTPs various feature attributes using DE-BP model, SiTime-GAN model and DPNG-EPMC ensemble clustering algorithm with modulation of microbial ecosystem health
Prediction-Clustering-Generation Tool Usage Guide:

I. For datasets of the same format in the provided example, including preprocessed input data (Waterplant) and corresponding output data on microbial community structure at various taxonomic levels (Phylum, Class, Order), the program can automatically perform model training and validation predictions. It outputs plots of training loss changes and total algorithmic error for both the BPNN model and DE-BP model, along with schematic diagrams comparing predicted microbial community compositions with original data.

1. Install anaconda，then run "conda install numpy / pandas / scipy/torch / tensorflow/matplotlib/streamlit / flask / fastapi/sklearn" and so on；
2. Open the "Anaconda prompt" program, then switch to the folder 'Prediction, Generation of WWTPs microbiome community structures' containing web_app.py；
3. All example standard csv format files are stored in folder 'Prediction, Generation of WWTPs microbiome community structures'；
4. Run the command "python -m streamlit run web_app.py"；
5. Enter relevant parameters, file paths, and the test instance ID after model training. Click Run to execute；
6. Outputs will include result.txt, Gabp_log.xlsx, Test_instance_output.csv, BP neural network loss curve charts, and line charts comparing predicted and actual values for each dimension of the testset，all output files are downloadable.

II. For datasets of the same format in the provided example, including preprocessed input data (Waterplant) and corresponding output data on microbial community structure (additionally added water plant and microbial data generated by SiTime-GAN model) at various taxonomic levels (Phylum, Class, Order), the program can automatically perform model training and validation predictions. It outputs plots of training loss changes and total algorithmic error for both the BPNN model and DE-BP model, along with schematic diagrams comparing predicted microbial community compositions with original data.

1. Install anaconda，then run "conda install numpy / pandas / scipy/torch / tensorflow/matplotlib/streamlit / flask / fastapi/sklearn" and so on；
2. Open the "Anaconda prompt" program, then switch to the folder 'Prediction, Generation of WWTPs microbiome community structures' containing web_app_generative.py；
3. All example standard csv format files are stored in folder 'Prediction, Generation of WWTPs microbiome community structures'；
4. Run the command "python -m streamlit run web_app_generative.py"；
5. Enter relevant parameters, file paths, and the test instance ID after model training. Click Run to execute；
6. Outputs will include result_Generative.txt, Gabp_log(Generative).xlsx, Test_instance_output(Generative).csv, BP neural network loss curve charts, and line charts comparing predicted and actual values for each dimension of the testset, all output files are downloadable.


III. For datasets of the same format in the provided example，the clustering analysis tasks categorize water plant characteristics into four different feature attributes: environmental and geographical features, influent features, effluent features and reactor operational features.  The program can automatically perform clustering and analysis，and output the Information of each centroid, the WWTP ranking information of each centroid and the plot of Microbio heatmap under four different feature attributes at three taxonomic levels (Phylum, Class, Order). 

1. Open the files in MATLAB software，then run the  'web_app.m' file;
2.  All example standard csv format files are stored in subfolder 'clustering' of folder 'Clustering of WWTPs various feature attributes'；
2. Enter relevant parameters, Input file paths, Microbial_hierarchy (Phylum, Class, Order,...), Feature_attribute (four different attributes...), Microbial_feature_number (21,51,171,...) and Output_folder. Click Run to execute；
3. Outputs will include Information of each centroid.xlsx, the WWTP ranking information of each centroid.xlsx and Microbio heatmap.jpg, all output files are downloadable.

